### PR TITLE
[genesis] Mainnet genesis tooling

### PR DESCRIPTION
**Pull request by gregnazario**
### Description
This adds mainnet genesis customization to the tooling.  Additionally, it handles lots of checks to make sure that everything is as it is expected to be.

### Test Plan
Tested manually against a detailed example

- - -
This change is [Reviewable](https://reviewable.io/reviews/aptos-labs/aptos-core/4613)

### CODE REVIEW

**Review by movekevin — APPROVED** [review]

File: crates/aptos-genesis/src/config.rs [hunk 1]
```diff
@@ -344,3 +354,217 @@ pub struct StringOperatorConfiguration {
...
+            // Check vesting schedule adds up properly
+            let mut numerators = 0;
+            let denominator = pool.vesting_schedule_denominator;
+            let mut last_numerator = 0;
+            for numerator in pool.vesting_schedule_numerators.iter() {
+                numerators += *numerator;
+                last_numerator = *numerator;
+            }
+
+            if numerators > denominator {
```
Comment (movekevin) [inline crates/aptos-genesis/src/config.rs:489]: Love it!

File: crates/aptos-genesis/src/config.rs [hunk 2]
```diff
@@ -344,3 +354,217 @@ pub struct StringOperatorConfiguration {
...
+                    "Numerators {} add up over the denominator {} for pool #{}",
+                    numerators,
+                    denominator,
+                    i
+                )
+            } else if (denominator - numerators) % last_numerator != 0 {
+                anyhow::bail!("Numerators don't add up to the denominator {} (with the last one {} being repeated for pool #{}", denominator, last_numerator, i)
+            }
+
+            // I'm going to assume no one wants to pay more than 50% of their rewards away
```
Comment (movekevin) [inline crates/aptos-genesis/src/config.rs:500]: Makes sense

File: crates/aptos-genesis/src/config.rs [hunk 3]
```diff
@@ -344,3 +354,217 @@ pub struct StringOperatorConfiguration {
...
+                if pool.validator.validator_host.is_none() {
+                    anyhow::bail!(
+                        "Pool #{} is setup to join during genesis but missing a validator host",
+                        i
+                    );
+                }
+                if pool.validator.validator_network_public_key.is_none() {
+                    anyhow::bail!("Pool #{} is setup to join during genesis but missing a validator network public key", i);
+                }
+                if pool.validator.stake_amount < 100000000000 {
```
Comment (movekevin) [inline crates/aptos-genesis/src/config.rs:526]: Make this 1M APT? This code path is only for mainnet anyway

File: crates/aptos/src/genesis/mod.rs [hunk 4]
```diff
@@ -119,10 +125,88 @@ pub fn fetch_mainnet_genesis_info(git_options: GitOptions) -> CliTypedResult<Mai
...
+    let employee_validators: Vec<_> = employee_vesting_accounts
+        .inner
+        .iter()
+        .map(|inner| inner.validator.clone())
+        .collect();
+    let employee_vesting_accounts: Vec<EmployeePool> = employee_vesting_accounts.try_into()?;
     let validators = get_validator_configs(&client, &layout, true).map_err(parse_error)?;
+
+    // Check accounts for employee accounts
+    for (i, pool) in employee_vesting_accounts.iter().enumerate() {
```
Comment (movekevin) [inline crates/aptos/src/genesis/mod.rs:162]: Nice!

**Review by movekevin — APPROVED** [review]

File: aptos-move/vm-genesis/src/lib.rs [hunk 5]
```diff
@@ -80,8 +80,8 @@ pub static GENESIS_KEYPAIR: Lazy<(Ed25519PrivateKey, Ed25519PublicKey)> = Lazy::
 });
 
 pub fn encode_aptos_mainnet_genesis_transaction(
-    accounts: &[AccountMap],
-    employees: &[EmployeeAccountMap],
+    accounts: &[AccountBalance],
```
Comment (banool) [inline aptos-move/vm-genesis/src/lib.rs:83]: nit: account_balances

File: crates/aptos-genesis/src/config.rs [hunk 6]
```diff
@@ -344,3 +354,217 @@ pub struct StringOperatorConfiguration {
...
+            }
+
+            if numerators > denominator {
+                anyhow::bail!(
+                    "Numerators {} add up over the denominator {} for pool #{}",
+                    numerators,
+                    denominator,
+                    i
+                )
+            } else if (denominator - numerators) % last_numerator != 0 {
```
Comment (banool) [inline crates/aptos-genesis/src/config.rs:496]: While we're at it should we check that the denominator isn't zero? (if we don't already)

**Review by banool — APPROVED** [review]
Looks good on first pass but I'll take another look later (you can land), this is important.

**Pull request merged**